01.md

```diff
@@ -43,9 +42,57 @@ To obtain the `event.id`, we `sha256` the serialized event. The serialization is
+  ...
+}
+```
+
+The first element of the tag array is referred to as the tag _name_ or _key_ and the second as the tag _value_. So we can safely say that the event above has an `e` tag set to `"5c83da77af1dec6d7289834998ad7aafbd9e2191396d75ec3cc27f5a77226f36"`, an `alt` tag set to `"reply"` and so on. All elements after the second do not have a conventional name.
+
+This NIP defines 3 standard tags that can be used accross all event kinds with the same meaning. They are as follows:
+
+- The `e` tag, used to refer to an event: `["e", <32-bytes hex of the id of another event>, <recommended relay URL, optional>]`
+- The `p` tag, used to refer to another user: `["p", <32-bytes hex of a pubkey>, <recommended relay URL, optional>]`
+- The `a` tag, used to refer to a parameterized replaceable event: `["a", <kind integer>:<32-bytes hex of a pubkey>:<d tag value>, <recommended relay URL, optional>]`
+
+As a convention, all single-letter key tags are expected to be indexed by relays, such that it is possible, for example, to query or subscribe to events that reference the event `"5c83da77af1dec6d7289834998ad7aafbd9e2191396d75ec3cc27f5a77226f36"` by using the `{"#e": "5c83da77af1dec6d7289834998ad7aafbd9e2191396d75ec3cc27f5a77226f36"}` filter.
+
+### Kinds
+
+Kinds specify how clients should interpret the meaning of each event and the other fields of each event (e.g. an `"r"` tag may have a meaning in an event of kind 1 and an entirely different meaning in an event of kind 10002). Each NIP may define the meaning of a set of kinds that weren't defined elsewhere. This NIP defines two basic kinds:
+
+- `0`: **metadata**: the `content` is set to a stringified JSON object `{name: <username>, about: <string>, picture: <url, string>}` describing the user who created the event. A relay may delete older events once it gets a new one for the same pubkey.
+- `1`: **text note**: the `content` is set to the **plaintext** content of a note (anything the user wants to say). Content that must be parsed, such as Markdown and HTML, should not be used. Clients should also not parse content as those.
+
+And also a convention for kind ranges that allow for easier experimentation and flexibility of relay implementation:
+
+- for kind `n` such that `1000 <= n < 10000`, events are **regular**, which means they're all expected to be stored by relays.
+- for kind `n` such that `10000 <= n < 20000`, events are **replaceable**, which means that, for each combination of `pubkey` and `kind`, only the latest event is expected to be stored by relays, older versions are expected to be discarded.
+- for kind `n` such that `20000 <= n < 30000`, events are **ephemeral**, which means they are not expected to be stored by relays.
+- for kind `n` such that `30000 <= n < 40000`, events are **parameterized replaceable**, which means that, for each combination of `pubkey`, `kind` and the `d` tag, only the latest event is expected to be stored by relays, older versions are expected to be discarded.
```

Comment:
I want to know what @arthurfranca thinks about this one.

Reply:
I don't think there is a better place for them. The ranges are pretty much part of the core protocol now, pretty simple to understand and very important to know.

Reply:
Well, if we are to reorganize kinds, I'd like to suggest what bothers me from day 1 :)
What about making default kinds ranges limited to `32767` ? It's `40000` so not a big difference. And it's easily extendable with future NIPs. But at this point it would allow us to have 2 bytes to store kind instead of 4

Maybe I'm missing something implementation-wise here but I wasn't able to find an easy workaround to store currently existing kinds in 2-byte-way in postgres

Reply:
@alexgleason [my idea](https://github.com/nostr-protocol/nips/pull/683#issuecomment-1652239180) was to mark some NIPs as "required" exactly how @ariard said above: "Personally, I would advocate to update all the integrated NIPs (12, 16, 20 and 33) and add the mention “mandatory” and update the front README.md with the list of mandatory NIPs." But you can check on the previous link what the response was, at least it was longer than [this one](https://github.com/nostr-protocol/nips/pull/696#issuecomment-1667879114).

Now I keep my opinions to myself until something like [nips over nostr](https://github.com/nostr-protocol/nips/pull/704) ever becomes a reality.

Reply:
Also, isn't kind 3 a replaceable event? And kind 0.